### DISCUSSION TITLE
Run "main.py" directly without "boot.py"

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#12719] [General]
Hello everyone.

I have an application than run once "boot.py" code to save some parameters to NVM (non-volatile memory) as a "factory defaults".
Then the same "boot.py" renames itself to "no_boot.py" and then the execution flow tries to execute "main.py" but doesn't work.

However, If I click the "boot switch button" on my ESP32 Devkit v.1, then the "main.py" continues to executing the program correctly.

Are there some way to avoid this?
In other words, if we turn power on our device and "boot.py" doesn't exist, the microcontroller follows the main execution without stopping all program flow?

Thank you very much. Greetings

**Comment by robert-hh** [comment]
The normal habit of MP is to run boot.py, if it exists, and then main.py, if it exists. So if your boot.py renames itself and then terminates, main.py will be executed.
Edit: if boot.py fails with an exception, main.py is skipped.

**Reply by nastiliano** [reply]
Ok,
I knew it. I read some time that it worked in this way... but is very curious, my code is simple and all is executed until the last line (the renaming line) and then fails...

**Reply by robert-hh** [reply]
May it be that the rename fails, e.g
 because the target exists? Try to put the rename into a try/except clause.

**Reply by nastiliano** [reply]
Ok, you were completly right...

My code failed in "NVS" lines... "NVS" is hard to understand.
I'm going to open a new post to ask my doubts about Non Volatile Storage.

Thank you very much. Best regards.

**Comment by EDIandXML** [comment]
Here's my solution of this main.py problem.
And the solution is to short-circuit the program before it really gets going.
Here's the short circuit code, using pins 0 and 1.  Change the pin numbers appropriate to your situation. Place the code near the top of your progam.
```
import sys
doWeRunPin0 = Pin(0, Pin.OUT)
doWeRunPin1 = Pin(1, Pin.IN)
doWeRunPin0.value(True)
if doWeRunPin1.value() == False:
    sys.exit() # we're done
```

If you have two GPIO pins you're not using then connect to the two pins when you want main.py to run.  Disconnect when you don't.

**Reply by EDIandXML** [reply]
I forgot 
at the beginning
`from machine import Pin`
and the end
`doWeRunPin0.value(False)`
